test/distributed/org/apache/cassandra/distributed/test/ReadRepairQueryTypesTest.java

```diff
@@ -0,0 +1,1215 @@
+            String columnsQuery = String.format("SELECT %s FROM %s %s", columns, tableName, restriction);
+            assertRows(executeDistributed(columnsQuery), columnsQueryResults);
+
+            return verifyQuery(allColumnsQuery, node1Rows, node2Rows);
+        }
+
+        /**
+         * Executes the specified column deletion on just one node. Then it runs the tested query with CL=ALL selectig
+         * only the specified columns (which are expected to be different to the deleted one) and verifies that it
+         * returns the specified rows. Then it runs the tested query again, this time selecting all the columns, to
+         * verify that the previous query didn't propagate the column deletion.
+         */
+        Tester deleteColumn(String columnDeletion,
+                            String columns,
+                            Object[][] columnsQueryResults,
+                            Object[][] node1Rows,
+                            Object[][] node2Rows)
+        {
+            assert restriction != null;
+
+            // execute the column deletion on just one node
+            populateNode(columnDeletion);
+
+            // execute the columns read with CL=ALL, in most cases this won't propagate the previous column deletion if
+            // the deleted and read columns don't overlap
+            String columnsQuery = String.format("SELECT %s FROM %s %s", columns, tableName, restriction);
+            assertRows(executeDistributed(columnsQuery), columnsQueryResults);
+
+            // verify the query selecting all the columns
+            return verifyQuery(allColumnsQuery, node1Rows, node2Rows);
+        }
+
+        /**
+         * Executes the specified row deletion on just one node and verifies the tested query, to ensure that the tested
+         * query propagates the row deletion.
+         */
+        Tester deleteRows(String rowDeletion, Object[][] node1Rows, Object[][] node2Rows)
+        {
+            populateNode(rowDeletion);
+            return verifyQuery(allColumnsQuery, node1Rows, node2Rows);
+        }
+
+        /**
+         * Verifies that the replicas are empty and drop the table.
+         */
+        void tearDown()
+        {
+            tearDown(rows(), rows());
+        }
+
+        /**
+         * Verifies the final status of the nodes with an unrestricted query, to ensure that the main tested query
+         * hasn't triggered any unexpected repairs. Finally, it drops the table.
+         */
+        void tearDown(Object[][] node1Rows, Object[][] node2Rows)
+        {
+            verifyQuery("SELECT * FROM " + tableName, node1Rows, node2Rows);
+            schemaChange("DROP TABLE " + tableName);
+        }
+
+        private Tester verifyQuery(String query, Object[][] node1Rows, Object[][] node2Rows)
+        {
+            // verify the per-replica status before running the query distributedly
+            assertRows(cluster.get(1).executeInternal(query), node1Rows);
```

Comment:
For future debuggers, it might be helpful to inject more explicit failure messages. These tests have multiple stages, so the more information we have on failure, the easier it will be to make sense of.

Reply:
Indeed it would be nice, but `AssertUtils.assertRows` doesn't accept an error message. If think we would need a new dtest API version to change this, right? Not sure we want to do that now, although it would be useful for all the tests, not only the RR ones.

Reply:
Ah, yeah, that might be worth a separate Jira.